### ISSUE TITLE
Tracking issue for supporting Nginx

### ISSUE THREAD

**Issue by lrh2000** [asterinas/asterinas#960]
## Configurations

Version: nginx-1.26.1

<details>
<summary>Building Configuration</summary>

For simplifity, I disable everything but `http` and `http_autoindex_module` (_no_ `--without-http` and `--without-http_autoindex_module` in the following command line) and build a static binary:

```
./configure --with-cc-opt="-O2" --with-ld-opt="-static" --without-select_module --without-poll_module --without-quic_bpf_module --without-http_charset_module --without-http_gzip_module --without-http_ssi_module --without-http_userid_module --without-http_access_module --without-http_auth_basic_module --without-http_mirror_module --without-http_geo_module --without-http_map_module --without-http_split_clients_module --without-http_referer_module --without-http_rewrite_module --without-http_proxy_module --without-http_fastcgi_module --without-http_uwsgi_module --without-http_scgi_module --without-http_grpc_module --without-http_memcached_module --without-http_limit_conn_module --without-http_limit_req_module --without-http_empty_gif_module --without-http_browser_module --without-http_upstream_hash_module --without-http_upstream_ip_hash_module --without-http_upstream_least_conn_module --without-http_upstream_random_module --without-http_upstream_keepalive_module --without-http_upstream_zone_module --without-http-cache --without-mail_pop3_module --without-mail_imap_module --without-mail_smtp_module --without-stream_limit_conn_module --without-stream_access_module --without-stream_geo_module --without-stream_map_module --without-stream_split_clients_module --without-stream_return_module --without-stream_pass_module --without-stream_set_module --without-stream_upstream_hash_module --without-stream_upstream_least_conn_module --without-stream_upstream_random_module --without-stream_upstream_zone_module --without-pcre --without-pcre2
```
</details>

<details>
<summary>Running Configuration </summary>

 - The explicit specification of the user (`user root`) is required to avoid errors due to `getgrnam("nobody")`.
 - The specific IP address (`listen 10.0.2.15:8080`) is required because we cannot support listening at `0.0.0.0`.
   - https://github.com/asterinas/asterinas/issues/264
```
user root;

events {
}

http {
    include mime.types;

    server {
        listen 10.0.2.15:8080;

        location / {
            autoindex on;
        }
    }
}
```
</details>

## TODOs

Resolve the following problems to run Nginx:
 - [x] https://github.com/asterinas/asterinas/pull/959
 - [x] Support `epoll_ctl(EPOLL_CTL_ADD)` on UNIX sockets.
<details>
<summary>Solution</summary>

This can be solved by adding the `register_observer()` implementation for UNIX sockets.

The implementation itself would be straightforward, but UNIX sockets might require more code refactoring, as has been done for TCP/UDP sockets.

</details>

 - [x] Support `prctl(PR_SET_DUMPABLE)`.
   - Done in https://github.com/asterinas/asterinas/pull/1039
<details>
<summary>Workaround</summary>

In `objs/ngx_auto_config.h`, comment out the following code:
```c
// #ifndef NGX_HAVE_PR_SET_DUMPABLE
// #define NGX_HAVE_PR_SET_DUMPABLE  1
// #endif
```

</details>

 - [x] Support `pwrite()`.
<details>
<summary>Workaround</summary>

In `objs/ngx_auto_config.h`, comment out the following code:
```c
// #ifndef NGX_HAVE_PWRITE
// #define NGX_HAVE_PWRITE  1
// #endif
```

</details>

 - [x] Support `ioctl(FIONBIO)`.
   - Done in https://github.com/asterinas/asterinas/pull/1041
<details>
<summary>Workaround</summary>

In `objs/ngx_auto_config.h`, comment out the following code:
```c
// #ifndef NGX_HAVE_FIONBIO
// #define NGX_HAVE_FIONBIO  1
// #endif
```

</details>

 - [x] Support `ioctl(FIOASYNC)`.
   - Done in https://github.com/asterinas/asterinas/pull/1041
<details>
<summary>Workaround</summary>

In `src/os/unix/ngx_process.c`, comment out the following code:
```c
        // if (ioctl(ngx_processes[s].channel[0], FIOASYNC, &on) == -1) {
        //     ngx_log_error(NGX_LOG_ALERT, cycle->log, ngx_errno,
        //                   "ioctl(FIOASYNC) failed while spawning \"%s\"", name);
        //     ngx_close_channel(ngx_processes[s].channel, cycle->log);
        //     return NGX_INVALID_PID;
        // }
```

</details>

 - [x] Support `ioctl(F_SETOWN)`.
   - Done in https://github.com/asterinas/asterinas/pull/1040
<details>
<summary>Workaround</summary>

In `src/os/unix/ngx_process.c`, comment out the following code:
```c
        // if (fcntl(ngx_processes[s].channel[0], F_SETOWN, ngx_pid) == -1) {
        //     ngx_log_error(NGX_LOG_ALERT, cycle->log, ngx_errno,
        //                   "fcntl(F_SETOWN) failed while spawning \"%s\"", name);
        //     ngx_close_channel(ngx_processes[s].channel, cycle->log);
        //     return NGX_INVALID_PID;
        // }
```

</details>

Finally, add the benchmark:
 - [x] https://github.com/asterinas/asterinas/pull/1385

**Comment by StevenJiang1110** [comment]
FYI: I provided support for nginx last year in [this branch](https://github.com/StevenJiang1110/asterinas/tree/dev-nginx-new). However, that branch contains a significant amount of workaround code that was never merged into the main branch, and it may not be compatible with the latest main branch as considerable time has passed.

The nginx version I tested was built using the following commands.
```bash
wget https://nginx.org/download/nginx-1.25.3.tar.gz
tar xvf nginx-1.25.3.tar.gz
cd nginx-1.25.3
./configure --without-http_rewrite_module --without-http_gzip_module
make -j
make install
``` 

I also enable a http benchmark `wrk` on nginx. And the commands to build `wrk` are
```bash
cd /root
git clone git@github.com:wg/wrk.git
cd wrk
make
```
